### PR TITLE
chore: Adds CoudEvents when a ScaledObject scales a workload from/to zero or one

### LINKED ISSUE #3527: Provide CloudEvents when scaling workload(s) from/to 0
## Event Types

- [ ] keda.workload.scaletozero.v1
- [ ] keda.workload.scalefromzero.v1

## Payload:
```json
{
    "specversion" : "1.0",
    "type" : "keda.workload.scaletozero.v1",
    "subject" : "/{kubernetes-cluster-name}/{namespace}/workload/{resource-name}/",
    "time" : "2018-04-05T17:31:00Z",
    "datacontenttype" : "application/json",
    "data" : {
      "scaleTarget": {
         "name": "sample-workload",
         "apiVersion": "Deployment",
         "kind": "apps/v1",
      }
   }
}
```

### Use-Case

Allow end-users to get notified when workloads are scaling to/from 0.

### Anything else?

Relates to https://github.com/kedacore/keda/issues/479

### PR DESCRIPTION
**Pull request by neelanjan00**
### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to #3527

### CODE REVIEW

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 1]
```diff
@@ -272,6 +275,13 @@ func (e *scaleExecutor) scaleToZeroOrIdle(ctx context.Context, logger logr.Logge
 
 			e.recorder.Eventf(scaledObject, corev1.EventTypeNormal, eventreason.KEDAScaleTargetDeactivated,
 				"Deactivated %s %s/%s from %d to %d", scaledObject.Status.ScaleTargetKind, scaledObject.Namespace, scaledObject.Spec.ScaleTargetRef.Name, currentReplicas, scaleToReplicas)
+
+			if scaleToReplicas == 0 {
+				e.eventEmitter.Emit(scaledObject, types.NamespacedName{Namespace: scaledObject.Namespace},
+					corev1.EventTypeNormal, eventemitter.ScaleFromZeroType, eventreason.KEDAScaleTargetActivated,
+					message.ScaleTargetFromZero)
+			}
```
Comment (JorTurFer) [inline pkg/scaling/executor/scale_scaledobjects.go:283]: I think that this should be scale to zero, shouldn't?

Comment (neelanjan00) [reply pkg/scaling/executor/scale_scaledobjects.go]: You're right, fixed.

File: pkg/common/message/message.go [hunk 2]
```diff
@@ -28,4 +28,8 @@ const (
 	ScaleTargetNotFoundMsg = "Target resource doesn't exist"
 
 	ScaleTargetNoSubresourceMsg = "Target resource doesn't expose /scale subresource"
+
+	ScaleTargetFromZero = "Target resource is scaling up from zero number of replicas"
```
Comment (tomkerkhove) [inline pkg/common/message/message.go:32]:
```suggestion
	ScaleTargetFromZero = "Target resource is scaling out from zero"
```

File: pkg/common/message/message.go [hunk 3]
```diff
@@ -28,4 +28,8 @@ const (
...
+
+	ScaleTargetToZero = "Target resource is scaling down to zero number of replicas"
```
Comment (tomkerkhove) [inline pkg/common/message/message.go:34]:
```suggestion
	ScaleTargetToZero = "Target resource is scaling in to zero replicas"
```

**Comment by tomkerkhove** [comment]
Can you open a PR for our docs as well please?

**Review by tomkerkhove — COMMENTED** [review]
Added some small suggestions, but I was wondering if:
- You could add e2e tests
- Share an example of the cloudevents being emitted

**Comment by zroubalik** [comment]
@neelanjan00 any update on this please?

**Comment by SpiritZhou** [comment]
eventEmitter.Emit() will emit a k8s event at the same time. Replacing the recorder with eventEmitter and r.Recorder.Event() with eventEmitter.Emit() is better than passing an eventEmitter as a new parameter. Alternatively, we could pass an eventEmitter as a new parameter and delete the recorder once all r.Recorder.Event() calls have been replaced.

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 4]
```diff
@@ -272,6 +275,13 @@ func (e *scaleExecutor) scaleToZeroOrIdle(ctx context.Context, logger logr.Logge
 
 			e.recorder.Eventf(scaledObject, corev1.EventTypeNormal, eventreason.KEDAScaleTargetDeactivated,
```
Comment (SpiritZhou) [inline pkg/scaling/executor/scale_scaledobjects.go:276]: This one and the new one are duplicated. We should merge these two into one event. eventEmitter.Emit() includes emitting k8s event.

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 5]
```diff
@@ -309,6 +319,13 @@ func (e *scaleExecutor) scaleFromZeroOrIdle(ctx context.Context, logger logr.Log
...
 			"Original Replicas Count", currentReplicas,
 			"New Replicas Count", replicas)
+
+		if currentReplicas == 0 && e.eventEmitter != nil {
+			e.eventEmitter.Emit(scaledObject, types.NamespacedName{Namespace: scaledObject.Namespace},
+				corev1.EventTypeNormal, eventemitter.ScaleFromZeroType, eventreason.KEDAScaleTargetActivated,
+				message.ScaleTargetFromZero)
+		}
+
 		e.recorder.Eventf(scaledObject, corev1.EventTypeNormal, eventreason.KEDAScaleTargetActivated, "Scaled %s %s/%s from %d to %d", scaledObject.Status.ScaleTargetKind, scaledObject.Namespace, scaledObject.Spec.ScaleTargetRef.Name, currentReplicas, replicas)
```
Comment (SpiritZhou) [inline pkg/scaling/executor/scale_scaledobjects.go:329]: This one and the new one are duplicated. We should merge these two into one event.

File: controllers/keda/scaledjob_controller.go [hunk 6]
```diff
@@ -80,7 +80,7 @@ func init() {
 
 // SetupWithManager initializes the ScaledJobReconciler instance and starts a new controller managed by the passed Manager instance.
 func (r *ScaledJobReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
-	r.scaleHandler = scaling.NewScaleHandler(mgr.GetClient(), nil, mgr.GetScheme(), r.GlobalHTTPTimeout, mgr.GetEventRecorderFor("scale-handler"), r.SecretsLister)
+	r.scaleHandler = scaling.NewScaleHandler(mgr.GetClient(), nil, mgr.GetScheme(), r.GlobalHTTPTimeout, mgr.GetEventRecorderFor("scale-handler"), r.SecretsLister, nil)
```
Comment (SpiritZhou) [inline controllers/keda/scaledjob_controller.go:83]: create and pass a new event emitter as well so that it doesn't need to check nil later.

**Comment by tomkerkhove** [comment]
We are planning release on Thursday, any chance we can incorporate the feedback?

**Comment by neelanjan00** [comment]
Hey @tomkerkhove, @zroubalik sorry for not being active on this! I am presently occupied with some office work so not able to put time away for this, not sure if I will be able to add the E2E tests on time for the release.

**Comment by tomkerkhove** [comment]
No worries!

**Comment by tomkerkhove** [comment]
Any update on this @neelanjan00? If you prefer us to work on this then that is fine as well

**Comment by neelanjan00** [comment]
Hi @tomkerkhove sorry for being inactive on this, I don't think it will be feasible for me to work on the remaining tasks, it will be helpful if someone else can take it up. Thanks!

**Comment by tomkerkhove** [comment]
Closing as discussed on https://github.com/kedacore/keda/issues/3527#issuecomment-2206823001 @SpiritZhou is taking over and will use this one as inspiration - Thanks for working on this anyway @neelanjan00!

**Pull request closed without merging**